src/libcore/pin.rs

```diff
@@ -1,160 +1,284 @@
+    #[unstable(feature = "pin", issue = "49150")]
+    pub unsafe fn map_unchecked<U, F>(this: Pin<&'a T>, func: F) -> Pin<&'a U> where
+        F: FnOnce(&T) -> &U,
+    {
+        let pointer = &*this.pointer;
+        let new_pointer = func(pointer);
+        Pin::new_unchecked(new_pointer)
+    }
+
+    /// Get a safe reference out of a pin.
     #[unstable(feature = "pin", issue = "49150")]
-    pub fn reborrow<'b>(&'b mut self) -> PinMut<'b, T> {
-        PinMut { inner: self.inner }
+    pub fn get(this: Pin<&'a T>) -> &'a T {
+        this.pointer
     }
+}
 
-    /// Get a mutable reference to the data inside of this `PinMut`.
+impl<'a, T> Pin<&'a mut T> {
+    /// Get a mutable reference to the data inside of this `Pin`.
+    ///
+    /// # Safety
     ///
     /// This function is unsafe. You must guarantee that you will never move
     /// the data out of the mutable reference you receive when you call this
-    /// function.
+    /// function, so that the invariants on the `Pin` type can be upheld.
     #[unstable(feature = "pin", issue = "49150")]
-    pub unsafe fn get_mut_unchecked(this: PinMut<'a, T>) -> &'a mut T {
-        this.inner
+    pub unsafe fn get_mut_unchecked(this: Pin<&'a mut T>) -> &'a mut T {
+        this.pointer
     }
 
     /// Construct a new pin by mapping the interior value.
     ///
-    /// For example, if you  wanted to get a `PinMut` of a field of something,
+    /// For example, if you  wanted to get a `Pin` of a field of something,
     /// you could use this to get access to that field in one line of code.
     ///
+    /// # Safety
+    ///
     /// This function is unsafe. You must guarantee that the data you return
     /// will not move so long as the argument value does not move (for example,
     /// because it is one of the fields of that value), and also that you do
     /// not move out of the argument you receive to the interior function.
     #[unstable(feature = "pin", issue = "49150")]
-    pub unsafe fn map_unchecked<U, F>(this: PinMut<'a, T>, f: F) -> PinMut<'a, U> where
-        F: FnOnce(&mut T) -> &mut U
+    pub unsafe fn map_unchecked_mut<U, F>(this: Pin<&'a mut T>, func: F) -> Pin<&'a mut U> where
+        F: FnOnce(&mut T) -> &mut U,
     {
-        PinMut { inner: f(this.inner) }
-    }
-
-    /// Assign a new value to the memory behind the pinned reference.
-    #[unstable(feature = "pin", issue = "49150")]
-    pub fn set(this: PinMut<'a, T>, value: T)
-        where T: Sized,
-    {
-        *this.inner = value;
+        let pointer = Pin::get_mut_unchecked(this);
+        let new_pointer = func(pointer);
+        Pin::new_unchecked(new_pointer)
     }
 }
 
 #[unstable(feature = "pin", issue = "49150")]
-impl<'a, T: ?Sized> Deref for PinMut<'a, T> {
+impl<P, T> Deref for Pin<P> where
+    P: Deref<Target = T>,
+    T: ?Sized,
+{
     type Target = T;
-
     fn deref(&self) -> &T {
-        &*self.inner
+        &*self.pointer
```

Comment:
I'd prefer if you could use something like https://github.com/rust-lang/rust/issues/49150#issuecomment-416621688. In fact, I'd prefer if `Pin` could be in its own little private local submodule which only exposes those 4 functions (`as_ref`, `as_shr`, `unpin_mut`, `unpin_shr`) and `new_unchecked`; and everything else lived outside that module. Just to be sure we do not accidentally unpin something anywhere here.

Reply:
At the least, this could use `get` so that we only have one place witnessing the `Pin<&T> -> &T` conversion.

Reply:
Unfortunately, you can't quite implement it all just using `Deref` like that, since `Deref` only gives you the lifetime of the borrow, not the lifetime of the original object.

Reply:
Well there could be some core function that `Deref` and this uses.